src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -1679,27 +1679,22 @@ public void visitIf(JCIf tree) {
         @Override
         public void visitStringTemplate(JCStringTemplate tree) {
             JCExpression processor = tree.processor;
+            Type processMethodType = getProcessMethodType(tree, processor.type);
 
-            if (processor != null) {
-                scan(processor);
-                Type interfaceType = types.asSuper(processor.type, syms.processorType.tsym);
-
-                if (interfaceType != null) {
-                    List<Type> typeArguments = interfaceType.getTypeArguments();
-
-                    if (typeArguments.size() == 2) {
-                        Type throwType = typeArguments.tail.head;
-
-                        if (throwType != null) {
-                            markThrown(tree, throwType);
-                        }
-                    }
-                }
+            for (Type thrown : processMethodType.getThrownTypes()) {
+                markThrown(tree, thrown);
             }
 
             scan(tree.expressions);
         }
 
+        private Type getProcessMethodType(JCStringTemplate tree, Type processorType) {
```

Comment:
This method is basically the same as in `Attr`, correct? I wonder if there's a way to avoid the duplication? One possibility would be to compute the type in `Attr`, and store it in a field on `JCStringTemplate`, and just use it in `Flow`. Alternatively, it could be shared in some class, although only `TreeInfo` and `Resolve` come to mind, and neither seem particularly suitable for that.

Reply:
I thought of the TreeInfo and Resolve idea but the number of arguments required made it messy. Hanging off of JCStringTemplate seemed like clutter to me.

Reply:
Added